plugins/hls-cabal-gild-plugin/src/Ide/Plugin/CabalGild.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Ide.Plugin.CabalGild where
+
+import           Control.Monad.Except        (throwError)
+import           Control.Monad.IO.Class
+import qualified Data.Text                   as T
+import           Development.IDE             hiding (pluginHandlers)
+import           Ide.Plugin.Error            (PluginError (PluginInternalError, PluginInvalidParams))
+import           Ide.PluginUtils
+import           Ide.Types
+import           Language.LSP.Protocol.Types
+import           Prelude                     hiding (log)
+import           System.Directory
+import           System.Exit
+import           System.FilePath
+import           System.Process.ListLike
+import qualified System.Process.Text         as Process
+
+data Log
+  = LogProcessInvocationFailure Int T.Text
+  | LogReadCreateProcessInfo [String]
+  | LogInvalidInvocationInfo
+  | LogFormatterBinNotFound
+  deriving (Show)
+
+instance Pretty Log where
+  pretty = \case
+    LogProcessInvocationFailure exitCode err ->
+      vcat
+        [ "Invocation of cabal-gild failed with code" <+> pretty exitCode
+        , "Stderr:" <+> pretty err
+        ]
+    LogReadCreateProcessInfo args ->
+      "Formatter invocation: cabal-gild " <+> pretty args
+    LogInvalidInvocationInfo -> "Invocation of cabal-gild with range was called but is not supported."
+    LogFormatterBinNotFound -> "Couldn't find executable 'cabal-gild'"
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId =
+  (defaultCabalPluginDescriptor plId "Provides formatting of cabal files with cabal-gild")
+    { pluginHandlers = mkFormattingHandlers (provider recorder)
+    }
+
+-- | Formatter provider of cabal gild.
+-- Formats the given source in either a given Range or the whole Document.
+-- If the provider fails an error is returned that can be displayed to the user.
+provider :: Recorder (WithPriority Log) -> FormattingHandler IdeState
+provider recorder _ _ (FormatRange _) _ _ _ = do
+  logWith recorder Info LogInvalidInvocationInfo
+  throwError $ PluginInvalidParams "You cannot format a text-range using cabal-gild."
```

Comment:
You kind of can, depending on what exactly "range" means. For example:

``` sh
$ echo 'build-depends:{base>=4.19,bytestring>=0.12}' | cabal-gild
build-depends:
  base >=4.19,
  bytestring >=0.12
```

But I wonder if anyone would actually want this behavior anyway.

Reply:
I think people could find range formatting useful. But supporting only top-level stanzas and fields... Not sure it's the most intuitive to understand behaviour.

Reply:
Shall we make an issue so we can put a link to the issue in the error? :smiling_imp:

Reply:
I'd be happy to add support for formatting ranges if someone can point me to a resource that explains how it should work! As shown, Gild can format any subsection of a document that's still syntactically valid. But it can't, for example, format any arbitrary range. Like if you tried to format just the contents of the `build-depends` field, this is what you'd get:

``` sh
$ echo 'base>=4.19,bytestring>=0.12' | cabal-gild
base >= 4.19 , bytestring >= 0.12
```

Reply:
> I'd be happy to add support for formatting ranges if someone can point me to a resource that explains how it should work!

You mean "The document range formatting request is sent from the client to the server to format a given range in a document" isn't good enough for you? :joy: 

> As shown, Gild can format any subsection of a document that's still syntactically valid. But it can't, for example, format any arbitrary range.

Does it take into account the surrounding context? Perhaps it doesn't matter, but an easy way this kind of formatting could go wrong is if you format the range in isolation and therefore treat it as if it is at the top-level of the structure, which it might not be.

But otherwise I think it's fine if range formatting just doesn't do anything unless you've selected a range that includes a whole element that can be sensibly formatted.

Reply:
I think the only way it could take surrounding context into account is to maintain leading blank spaces. Otherwise it has no way of knowing that it's "inside" a `library` section, for example. However that doesn't matter too much for Gild; it doesn't keep track of that context anyway. The field names are unique enough that it just looks for `build-depends` rather than `build-depends` within `library`. See the documentation here: https://hackage.haskell.org/package/cabal-gild-1.1.0.0/docs/CabalGild-Action-FormatFields.html#v:parsers

How does range formatting work exactly? If Gild would only get the range's content as input, then there's not much it can do. But if it gets the entire buffer's input along with a span identifying the range to format, then potentially it could do something nice. (Although that might be a lot of work for me.)

Reply:
> But if it gets the entire buffer's input along with a span identifying the range to format,

That's what you get, indeed.

Reply:
I created an issue for this: https://github.com/tfausak/cabal-gild/issues/42. No idea if or when I'll get around to implementing it though.